File: FrontEnd/src/utils/transactionFilters.ts
import type { Transaction } from '../services/transactionsService'

export interface TransactionFilters {
  search: string
  account: string
  category: string
  needsCategoryOnly: boolean
}

// account/category are the empty string for "All accounts"/"All categories" (no filtering on
// that dimension) - matches the <select> sentinel option's value being "".
export function filterTransactions(transactions: Transaction[], filters: TransactionFilters): Transaction[] {
  const search = filters.search.trim().toLowerCase()

  return transactions.filter((t) => {
    if (search && !t.description.toLowerCase().includes(search)) {
      return false
    }
    if (filters.account && t.account !== filters.account) {
      return false
    }
    if (filters.category && t.category !== filters.category) {
      return false
    }
    if (filters.needsCategoryOnly && t.category) {
      return false
    }
    return true
  })
}

File: iosApp/iosApp/Utils/TransactionFilters.swift
import Foundation

// The search + needs-category predicates from FrontEnd/src/utils/transactionFilters.ts - the only
// two the iOS transactions screen exposes (the web's account / category / amount-sign / hide-ignored
// filters are not in the iOS mockup).
enum TransactionFilters {
    // Case-insensitive substring match on the description; an empty/whitespace query matches all.
    static func matchesSearch(_ transaction: Transaction, query: String) -> Bool {
        let search = query.trimmingCharacters(in: .whitespaces).lowercased()
        guard !search.isEmpty else { return true }
        return transaction.description.lowercased().contains(search)
    }

    // Ignored transactions never get (or need) a category, so they don't count as "needing" one.
    static func needsCategory(_ transaction: Transaction) -> Bool {
        transaction.isUncategorized && transaction.ignore != true
    }

    static func filter(_ transactions: [Transaction], search: String, needsCategoryOnly: Bool) -> [Transaction] {
        transactions.filter { transaction in
            guard matchesSearch(transaction, query: search) else { return false }
            if needsCategoryOnly, !needsCategory(transaction) { return false }
            return true
        }
    }

    static func needsCategoryCount(_ transactions: [Transaction]) -> Int {
        transactions.filter(needsCategory).count
    }
}
